### PR TITLE
enhancement-17 output directory parameter added, testing, and default…

### PR DESCRIPTION
**Pull request by benmcosker**
… option set to create piped output as well as a parameter to specify a user specified output directory.  Testing refactored to cover code from `gulp build`.  Old testing was not providing coverage.

@obuckley-kenzan / @paul-barry-kenzan / @jpoo90

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -48,8 +48,9 @@
   },
   "devDependencies": {
     "gulp": "3.9.0",
+    "gulp-clean": "^0.3.2",
```
Comment (obuckley-kenzan) [inline package.json]: please remove the `^`

**Comment by obuckley-kenzan** [comment]
ok to test

File: package.json [hunk 2]
```diff
@@ -48,8 +48,9 @@
...
+    "gulp-expect-file": "0.0.7",
     "pipeline-test-node": "0.4.0",
```
Comment (benmcosker) [inline package.json]: updating to add README.md edits.

**Comment by paul-barry-kenzan** [comment]
ok to test ?

**Comment by benmcosker** [comment]
@paul-barry-kenzan - ok to test.  I will update to use pipeline-handyman for file cleanup tasks in the testing.

File: package.json (hunk 1 above)
Comment (benmcosker) [inline package.json]: @obuckley-kenzan - this currently uses pipeline-handyman 0.3.0, which has no clean method.  Will see if updating to 1.0.0 is a working solution.

**Comment by obuckley-kenzan** [comment]
cloudbees is down, need to get back up and running

**Comment by benmcosker** [comment]
@obuckley-kenzan - This was tested using `gulp-build`, this is what the jenkins job will run against I presume. Correct?

**Comment by obuckley-kenzan** [comment]
ok to test

**Comment by benmcosker** [comment]
@obuckley-kenzan - cloudbees is still having issues. Job seems to be hung up since yesterday.

**Comment by obuckley-kenzan** [comment]
@paul-barry-kenzan 
Thoughts?

File: README.md [hunk 3]
```diff
@@ -44,10 +44,12 @@ Pipeline options:
 
     * __autoprefix__ If you don't want to have your CSS rules prefixed set this property to __false__.
     
-    * __concatCSS__ If set to __false__ the pipeline won't concatenate the files to generate a single CSS file.
+    * __concatCSS__ If set to __false__ the pipeline won't concatenate the files to generate a single CSS file and will concatenate them same directory as the original files..
```
Comment (obuckley-kenzan) [inline README.md]: A question here:
> and will concatenate them same directory as the original files.

if concat is set to false, then there should be no concatenation happening, right?  Or maybe the wording is incorrect?

File: README.md [hunk 4]
```diff
@@ -44,10 +44,12 @@ Pipeline options:
...
 
     * __outputFileName__ If __concatCSS__ is set to __true__, this value will be used to name the file. By default, the config is set to get the name of the package that consumes pipeline-compile-less. So for example, it would end up named as `your-project.css`. If you set this value, do not suffix the string with '.css', as this is handled internally.
 
+    * __outputDirectory__ If __concatCSS__ is set to __true__, . If you set this value, the .MAP and CSS files will be written to that directory. The default directory, if none is specified, is `dest`.
```
Comment (obuckley-kenzan) [inline README.md]: map files should be lowercase, eg.  `.map`

File: README.md [hunk 5]
```diff
@@ -82,7 +85,8 @@ This pipeline returns an object. This object receives a stream with the LESS fil
   
   + If _config.outputFileName_ is set, that name will be used (and suffixed with '.css') instead of picking up the consuming package name as a default value. For example, the compiled file would be named as `your-custom-value.css` if you used 'your-custom-value'.
 
-
+  + If _config.outputDirectory is set, that name will be used to output .MAP and .CSS files, provided _config.concatCSS_ is __true__.
```
Comment (obuckley-kenzan) [inline README.md]: same here with `.map`

File: README.md (hunk 3 above)
Comment (paul-barry-kenzan) [inline README.md]: Double period as well.

File: test/index.spec.js [hunk 6]
```diff
@@ -1,107 +1,104 @@
...
+        addSourceMaps: false
       }))
-      .pipe(assert.length(1))
-      .pipe(assert.first(function (file) {
-        var customPath = handyman.getPackageName() + '.css';
-
-        assertPath(file, customPath);
-      }));
-
+      .pipe(expect('dest/pipeline-compile-less.css'))
```
Comment (obuckley-kenzan) [inline test/index.spec.js]: does it make sense to assert no other files were generated either?

Comment (paul-barry-kenzan) [inline test/index.spec.js]: Agreed. I wouldn't suggest that for all cases, but in this instance, we're purposely only generating a single file, it would be nice to also check that ONLY a single file was created. Nice catch @obuckley-kenzan

File: test/index.spec.js [hunk 7]
```diff
@@ -1,107 +1,104 @@
...
 
+  it('Should output two files to default directory', function (done) {
+
+    gulp.src(getFixtures('*'))
+      .pipe(compilePipeline
+        .compileLESS({
+          autoprefix: false,
+          addSourceMaps: false,
+          concatCSS: false}))
+      .pipe(expect(['test/fixtures/test-less1.css', 'test/fixtures/test-less2.css']))
```
Comment (obuckley-kenzan) [inline test/index.spec.js]: @benmcosker 
Can you explain this test?  I don't think we should have tests generating files into the _src_ or _test_ (in other words version controlled directories) directories.

Also, should we add `alt` to _.gitignore_?

File: test/index.spec.js [hunk 8]
```diff
@@ -1,107 +1,104 @@
...
-var assert = require('stream-assert');
-var handyman = require('pipeline-handyman');
 var compilePipeline = require('../src/index.js');
+var clean = require('gulp-clean');
 var gulp = require('gulp');
 var path = require('path');
-var testPath = path.join(__dirname, 'dest');
+var expect = require('gulp-expect-file');
+
+var altPath = 'alt';
```
Comment (obuckley-kenzan) [inline test/index.spec.js]: I think we should refine the name a bit more.
`var altOutputPath = 'alt-output-directory/`

File: test/index.spec.js (hunk 7 above)
Comment (benmcosker) [inline test/index.spec.js]: @obuckley-kenzan - currently testing is the blocker as the files correctly generate on gulp build.  I am working to refactor this so that all of the files will both use gulp-handyman (and remove gulp-clean) AND keep all generated files to either the dest directory or the test/fixtures directory.  As it stands now, yes, the `alt` directory should be added to `.gitignore`

**Comment by paul-barry-kenzan** [comment]
@benmcosker 
I've done some digging here and found that the latest version of the test pipeline is your best bet. So, go ahead and update from `0.4.0` to `1.0.0`. This will require a bit of config to get a standard output. Pulling from the gulpfile of pipeline-validate-js, you can see the config object being passed into the pipeline method. 

https://github.com/kenzanlabs/pipeline-validate-js/blob/master/gulpfile.js

Once the package.json is updated, and the config object created, the tests should run and pass properly. Let me know if you run into any roadblocks.

As for the tests, I verified locally, and they pass without issue. Nice work.

File: test/index.spec.js (hunk 7 above)
Comment (paul-barry-kenzan) [inline test/index.spec.js]: Do we need to add a "temp" folder to `.gitignore`, or should we just remove the directory at the end of the test suite? A simple `after()` block could remove it when finished.

Comment (obuckley-kenzan) [inline test/index.spec.js]: Yeah, `tmp` or `temp` is generally a convention alternate output directory names.  I think at minimum each test should run a clean first.

**Comment by benmcosker** [comment]
still looking at using gulp-handyman to perform cleaning tasks for tests.

**Comment by benmcosker** [comment]
@obuckley-kenzan / @paul-barry-kenzan 

1) I have not found a method to pass the tests using the gulp-handyman (1.0.0), so my PR with gulp-clean stands as is:
`var handman = gulp.require('gulp-handyman')`
`beforeEach(function() {
  handyman.clean('dest');
});
`
2) the other suggested improvement to test that only one file is being written for with concatenation/no source maps.  This will require some sort of file system plugin `gulp-count/gulp-debug` and perhaps another assertion plugin such as `should`.  Thoughts on this before moving forward? I don't want to add unnecessary baggage but want to keep all features tested.

**Comment by paul-barry-kenzan** [comment]
Thanks for the update @benmcosker 

1. Are you referencing `pipeline-handyman` or are you planning on using `gulp-handyman` in its place?
2. Can you expand on why we'd need an additional plugin to assert that a file has *not* been created?

**Comment by benmcosker** [comment]
I am looking for a better solution than `gulp-expect-file`, thinking this one plugin: gulp-filelog
https://github.com/adamayres/gulp-filelog

might be more whats needed as it can both count and verify output.  In the tests for that plugin `should ` is used for the testing.  I am checking but this seems to be a cleaner test than the assert on `gulp-assert-stream` used in the current master of `pipeline-compile-less`.

**Comment by benmcosker** [comment]
@paul-barry-kenzan - according to:
https://github.com/kotas/gulp-expect-file#optionsreportunexpected

- the test would throw an error if any more than the one file specified writes to the directory:
https://github.com/benmcosker/pipeline-compile-less/blob/713702eb6843b04d8c380a798b538652c84d0b9e/test/index.spec.js#L29

- so the test is also testing that only one file is being written.  Any additional files would throw an error.

File: gulpfile.js [hunk 9]
```diff
@@ -13,15 +13,28 @@ var config = {
...
-  ]
+  ],
+
+  plugins: {
+    mocha: {
+      reporter: 'spec'
+    },
+    istanbul: {
+      includeUntested: true,
+      reporters: ['text-summary'],
```
Comment (obuckley-kenzan) [inline gulpfile.js]: @benmcosker 
it seems that this may be causing the build failure?  i don't think reports are getting generated for Jenkins to find.

File: test/index.spec.js [hunk 10]
```diff
@@ -1,107 +1,106 @@
...
-        addSourceMaps: false,
-        output: testPath,
-        concatCSS: false
-      }))
-      .pipe(assert.length(1))
-      .pipe(assert.first(function (file) {
-        assertPath(file, DEFAULT_PATH);
-      }));
-
+        .compileLESS({
```
Comment (obuckley-kenzan) [inline test/index.spec.js]: indentation looks off here

File: gulpfile.js (hunk 9 above)
Comment (benmcosker) [inline gulpfile.js]: @obuckley-kenzan - this is the same reporter settings as from pipeline-validate-js.  Do we use the Cobertura plugin and Istanbul?

Reviewing this 
http://www.mikkolehtinen.com/blog/2013-01-31-jenkins-setup-for-a-javascript-project/
- reviewing this, the defaults  are 'text-summary' and 'html'.  I can try adding html.  Do you know of a recommended setting?

Comment (benmcosker) [inline gulpfile.js]: this is my error in Jenkins:
```
[htmlpublisher] Archiving HTML reports...
[htmlpublisher] Archiving at PROJECT level /scratch/jenkins/workspace/CI-pipeline-compile-less/coverage/lcov-report to /home/kenzan/hudson_home/jobs/CI-pipeline-compile-less/htmlreports/Coverage_Report
ERROR: Specified HTML directory '/scratch/jenkins/workspace/CI-pipeline-compile-less/coverage/lcov-report' does not exist.
Build step 'Publish HTML reports' changed build result to FAILURE
Recording test results
ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?
```
directory: coverage/lcov-report  is generated on run.

**Comment by paul-barry-kenzan** [comment]
@benmcosker @obuckley-kenzan 
Jenkins is failing due to:

>ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?

**Comment by obuckley-kenzan** [comment]
I see the issue.  It seems 
1) Pipeline validate JS is not publishing reports
2) The node testing pipeline needs to be changed to output to reports, and by default it publishes to coverage

Seems like I need to update both jobs, but coverage for now is the intended output directory for test results for the time being

**Comment by obuckley-kenzan** [comment]
@paul-barry-kenzan

**Comment by paul-barry-kenzan** [comment]
@obuckley-kenzan @benmcosker 
Is `pipeline-validate-js` suppose to create a hard-copy report? That's not typical of linting. Usually just console based. The coverage reporting is done by `pipeline-test-node`, and should output to a default directory.

**Comment by benmcosker** [comment]
Verified that this is now functioning as per spec by pulling:
https://github.com/benmcosker/pipeline-compile-less 

- files concatenate
- files map
- concatenation can be turned off
- alternate directories can be used for output
- alternate filenames can be used for output

**Comment by obuckley-kenzan** [comment]
@paul-barry-kenzan 
yeah, sorry I meant the jenkins job.  Given our recent advances around pipeline-test-node, all my comments here related to that topic can / should be ignored.

File: gulpfile.js [hunk 11]
```diff
@@ -2,32 +2,49 @@
 
 var gulp = require('gulp');
 var compilePipeline = require('./src/index.js');
-var testPipeline = require('pipeline-test-node')();
-var validatePipeline = require('pipeline-validate-js')();
+
+var testPipeline = require('pipeline-test-node')({ plugins: {
```
Comment (obuckley-kenzan) [inline gulpfile.js]: @benmcosker 
If we update to test-node 1.0.1, you should be able to avoid having to provide all this boilerplate and just use the out of the box options.

**Comment by obuckley-kenzan** [comment]
@benmcosker 
If we update to pipeline-test-node@1.0.1, all the issues cited in my comments regarding outputting directories and the such should be resolved.

**Comment by obuckley-kenzan** [comment]
@paul-barry-kenzan / @benmcosker 
Looks like we might have to set the threshold lower here in the configuration.  Lowest score is for branches at 75% coverage.  We can start with that and try and enforce progressive coverage through Jenkins.

**Comment by obuckley-kenzan** [comment]
fixed jenkins config

**Comment by obuckley-kenzan** [comment]
Seems like that threshold config should work.  See if you can debug this issue locally.

**Comment by benmcosker** [comment]
@obuckley-kenzan - will do. I don't need to keep hitting Jenkins with this failure.

**Comment by benmcosker** [comment]
@obuckley-kenzan - so it looks like passing a value to the istanbul through plugins is not working, though (locally) if I change the coverage from 80 to 70, I am able to pass coverage.   Looks to be an issue I will have a PR together shortly for so that https://github.com/kenzanlabs/pipeline-test-node/blob/master/src/index.js#L29 is able to be a configurable parameter.

**Comment by obuckley-kenzan** [comment]
@benmcosker 
Can you see what happens if you change this line locally within your _node_modules_ so it passes the `threshold` property to `istanbul.enforceThresholds`
https://github.com/kenzanlabs/pipeline-test-node/blob/master/src/index.js#L53
```javascript
.pipe(istanbul.enforceThresholds, config.plugins.istanbul.threshold);
```

**Comment by benmcosker** [comment]
passing this in the pipeline-compile-less/gulpfile.js `.pipe(testPipeline.test(plugins));` :
```var plugins = {
  istanbul: {
    thresholds: {
      global: 70
    }
  }};
```
and putting `config.plugins.istanbul.threshold` gets this message:

```Message:
    Cannot read property 'global' of undefined
Details:
    domain: [object Object]
    domainThrown: true
```

I can console out from within the pipeline factory in `pipeline-test-node` and returns the correct/passed through value, but the above error occurs when passed into  `istanbul.enforceThresholds`. Looking further into documentation to see whats expected.

I also tried wrapping the plugins in the `pipeline-compile-less` as such:

```
var options = {
  plugins: {
    istanbul: {
      threshold: {
        global: 70
      }
    }
  }
};
```

File: .editorconfig [hunk 12]
```diff
@@ -1,12 +0,0 @@
-root = true
```
Comment (obuckley-kenzan) [inline .editorconfig]: why was this file deleted?

**Comment by obuckley-kenzan** [comment]
re-running since `npm install` failed

**Comment by obuckley-kenzan** [comment]
looks like there are some issues with npmjs serving tarballs, which is causing the build to fail on `npm install`
http://status.npmjs.org/

will have to try again in a little while

**Comment by benmcosker** [comment]
@obuckley-kenzan - thats fine, this can wait till npmjs is working better. I made one last change and pushed to just lower the testing threshold on just branches.  Passes locally.

**Comment by obuckley-kenzan** [comment]
systems reporting as operational, trying again

**Comment by obuckley-kenzan** [comment]
@benmcosker 
success

File: test/index.spec.js [hunk 13]
```diff
@@ -1,107 +1,101 @@
...
+    clean({force: true});
+  });
+
+  it('Should output two files to default directory', function (done) {
 
+    gulp.src(getFixtures('*'))
+      .pipe(compilePipeline.compileLESS({
+        autoprefix: false,
+        addSourceMaps: false,
+        concatCSS: false}))
```
Comment (obuckley-kenzan) [inline test/index.spec.js]: can we wrap the the `}))` on the next line for readability?

File: test/index.spec.js [hunk 14]
```diff
@@ -1,107 +1,101 @@
...
-      }));
+describe('output to another filename', function() {
+
+  it('Should two files directory with specified filename', function (done) {
 
+    gulp.src(getFixtures('*'))
+      .pipe(compilePipeline.compileLESS({
+        outputFileName: 'test-filename.css',
+        outputDirectory: altOutputPath,
+        output: altOutputPath}))
```
Comment (obuckley-kenzan) [inline test/index.spec.js]: same here with wrapping `}))`

**Comment by obuckley-kenzan** [comment]
@paul-barry-kenzan / @benmcosker 
This seems good but I feel like we may be missing some use cases in the testing.  If we could just do a quick audit of the README options and track anything that's unaccounted for in an issue, I am ok to let this go in.

**Comment by benmcosker** [comment]
@paul-barry-kenzan / @obuckley-kenzan  - all I am seeing as untested (though it is passed though as `false`) is the `autoprefix` option.  Would a good way to test to see if this is working to use a known CSS tag that is autoprefixed like `border-radius`?  The CSS file would have to be checked that the values are present by reading the generated CSS file for rewritten/autoprefixed CSS... correct?

**Comment by obuckley-kenzan** [comment]
I think in general, going forward for pipelines, we should make sure the tests follow closely to the config options.  So in this case we would have describe blocks for each option.  This will make it easier to parse each set of tests.

If we want to make that part of #21 I'm ok with that.

**Comment by benmcosker** [comment]
@obuckley-kenzan - agreed.  Although the I was testing the autoprefix as false (and defaulting it to true) I was not checking the actual contents. Is checking the contents for autoprefixing getting into the area where I am checking another plugins function so not considered testworthy? I had thought so, but a strict reading of the README.md made me question that.

**Comment by obuckley-kenzan** [comment]
I think we should continue this conversation in the other ticket, if we want to expand upon our testing strategy.  For now I would like like to get a consensus on the original intent, which was to fix the bug
